spyder/plugins/ipythonconsole/widgets/main_widget.py

```diff
@@ -1432,203 +1357,114 @@ def create_new_client(self, give_focus=True, filename='', is_cython=False,
-                    stdin_port=kernel_client.stdin_port,
-                    hb_port=kernel_client.hb_port,
-                    control_port=kernel_client.control_port)
-                newports = self.tunnel_to_kernel(connection_info, hostname,
-                                                 sshkey, password)
-                (kernel_client.shell_port,
-                 kernel_client.iopub_port,
-                 kernel_client.stdin_port,
-                 kernel_client.hb_port,
-                 kernel_client.control_port) = newports
-            except Exception as e:
-                QMessageBox.critical(self, _('Connection error'),
-                                     _("Could not open ssh tunnel. The "
-                                       "error was:\n\n") + str(e))
-                return
+        client = ClientWidget(
+            self,
+            id_=client_id,
+            given_name=given_name,
+            config_options=self.config_options(),
+            additional_options=self.additional_options(),
+            interpreter_versions=self.interpreter_versions(),
+            context_menu_actions=self.context_menu_actions,
+            handlers=self.registered_spyder_kernel_handlers
+        )
 
-        # Assign kernel manager and client to shellwidget
-        kernel_client.start_channels()
-        shellwidget = client.shellwidget
+        # add hostname for get_name
+        client.hostname = hostname
 
-        if not known_spyder_kernel:
-            shellwidget.sig_is_spykernel.connect(
-                self.connect_external_spyder_kernel)
-        shellwidget.set_kernel_client_and_manager(
-            kernel_client, kernel_manager)
-        shellwidget.sig_exception_occurred.connect(
-            self.sig_exception_occurred)
-        self.sig_shellwidget_created.emit(shellwidget)
-        kernel_client.stopped_channels.connect(
-            lambda: self.sig_shellwidget_deleted.emit(shellwidget))
+        # Adding a new tab for the client
+        self.add_tab(client, name=client.get_name())
 
         # Set elapsed time, if possible
-        if not is_external_kernel:
-            self.set_client_elapsed_time(client)
+        if master_client is not None:
+            client.t0 = master_client.t0
+            client.timer.timeout.connect(client.show_time)
+            client.timer.start(1000)
 
-        # Adding a new tab for the client
-        self.add_tab(client, name=client.get_name())
+        try:
+            # Get new client for kernel
+            if master_client is not None:
+                kernel = master_client.kernel.copy()
+            else:
+                kernel = KernelConnection.from_connection_file(
+                    connection_file, hostname, sshkey, password)
+        except Exception as e:
+            client.show_kernel_error(e)
+            return
 
-        # Register client
-        self.register_client(client)
+        # Connect kernel
+        client.connect_kernel(kernel)
```

Comment:
I think clients hold now a reference to a `KernelConnection` object, right? If that's the case, then that `kernel` attribute should be renamed to `kernel_handler` or something else for clarity.